File: picasso-transformations/src/main/java/io/github/hendraanggrian/picassotransformations/CropRoundedTransformation.java
package io.github.hendraanggrian.picassotransformations;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;

import com.squareup.picasso.Transformation;

import io.github.hendraanggrian.picassotransformations.internal.PaintBuilder;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class CropRoundedTransformation implements Transformation {

    static final String TAG = "cropRounded(radius=%s, margin=%s)";

    private int radius;
    private int margin;

    CropRoundedTransformation(int radius, int margin) {
        this.radius = radius;
        this.margin = margin;
    }

    @Override
    public Bitmap transform(Bitmap source) {
        final int width = source.getWidth();
        final int height = source.getHeight();
        final Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);

        float right = width - margin;
        float bottom = height - margin;
        new Canvas(bitmap).drawRoundRect(new RectF(margin, margin, right, bottom), radius, radius, new PaintBuilder(Paint.ANTI_ALIAS_FLAG)
                .shader(source)
                .build());

        source.recycle();
        return bitmap;
    }

    @Override
    public String key() {
        return String.format(TAG, radius, margin);
    }
}

File: picasso-utils/src/com/hendraanggrian/picasso/transformation/CropRoundedTransformation.kt
package com.hendraanggrian.picasso.transformation

import android.graphics.*
import android.graphics.Bitmap.Config.ARGB_8888
import android.graphics.Bitmap.createBitmap
import android.graphics.Paint.ANTI_ALIAS_FLAG
import android.graphics.Shader.TileMode.CLAMP
import com.squareup.picasso.Transformation

internal class CropRoundedTransformation(private val radius: Int, private val margin: Int) : Transformation {

    override fun transform(source: Bitmap): Bitmap {
        val right = (source.width - margin).toFloat()
        val bottom = (source.height - margin).toFloat()
        val target = createBitmap(source.width, source.height, ARGB_8888)
        Canvas(target).drawRoundRect(RectF(margin.toFloat(), margin.toFloat(), right, bottom), radius.toFloat(), radius.toFloat(), Paint(ANTI_ALIAS_FLAG).apply {
            shader = BitmapShader(source, CLAMP, CLAMP)
        })
        source.recycle()
        return target
    }

    override fun key() = "CropRoundedTransformation[margin=$margin, radius=$radius]"
}